fix missing dependency links in Makefiles

vtjnash opened the pull request:
Make sure make knows how to re-build all outputs, and in the right orders.

vtjnash committed:
build: fix missing dependency links in Makefiles

staticfloat commented on cli/Makefile:
```diff
@@ -108,39 +108,38 @@ ABI_UNDERSCORE :=
+	@# Note that if the objcopy command starts getting too long, we can use `@file` to read
+	@# command-line options from `file` instead.
+	@$(call PRINT_LINK, $(OBJCOPY) $@.a $(STRIP_EXPORTED_FUNCS) $@.a.stripped)
+	@mv $@.a.stripped $@.a
```
It makes me slightly uncomfortable that these commands are being tied to the `.$(SHLIB_EXT)` variants.  I think it would be better to have these as part of the source build so that we can't get into a situation where someone accidentally deletes only the `.$(JL_MAJOR_MINOR_SHLIB_EXT)` file and then rebuilds and the `$(OBJCOPY)` stuff never happens, because the link already exists.

vtjnash replied:
yeah, I was editing this part blind, and CI shows I need to fix it

staticfloat commented on Makefile:
```diff
@@ -75,7 +75,7 @@ julia-libllvmcalltest: julia-deps
 	@$(MAKE) $(QUIET_MAKE) -C $(BUILDROOT)/src $*
 
-julia-cli-release julia-cli-debug: julia-cli-% :
+julia-cli-release julia-cli-debug: julia-cli-% : julia-deps
```
I think this is not strictly necessary.  I agree it feels a little weird for the builder to get built before anything else, but it _is_ pretty stand-alone....

vtjnash replied:
It has happened pretty often for me on my new Nanosoldier builder

vtjnash replied:
FWIW, the conflict happens because of the CompilerSupportLibraries, which change the linking

staticfloat replied:
What happens?  I don't think anything in `cli` depends on anything in `deps`, does it?

staticfloat replied:
Ahhh, on Linux and such we may actually link against `libstdc++` and whatnot.  I think we are less standalone on Linux than we are on Windows.

How do you feel about that?  Would you prefer that we are more consistent and are equally standalone on all platforms?  This was not 100% purposeful on my part, it was more that because we have RPATH it's not fatal to depend on these compiler support libraries on Linux/macOS, but we can certainly make things more consistent.

vtjnash replied:
It doesn't really matter, we just need to declare it correctly here

vtjnash committed:
build,uninstall: avoid removing directories in usr

This can cause a race if we happen to be trying to install something
else in the same directory we just removed. Very unlikely, but
unnecessary to remove them anyways (and generally was already broken for
Yggdrasil-provided binaries).
